src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    public class EnumerableToDispatchMarshaler : ICustomMarshaler
```

Comment:
Where is it manually used?

I have been working under assumption that these types are just internal implementation details, not referenced by user code: https://github.com/dotnet/corefx/pull/33773#issuecomment-443389274

Reply:
As of now it's used in the tests only. But it is public surface area and has to be manually used, at least for getting a managed `IEnumerable` from a native `IDispatch*`. That direction currently throws without the marshaler added with a `MarshalAs` attribute.

Reply:
So is the plan to also expose these types via CoreFX as well?

Reply:
> So is the plan to also expose these types via reference assemblies in CoreFX as well?

Ping ... I would be curious to understand the plan here.

Reply:
My initial thought would be to expose them via a CustomMarshalers.dll ref assembly in corefx. However, since they're so rarely used, I could see us just putting them into the System.Runtime.InteropServices ref assembly. For now I'm thinking we just keep them internal and have the tests for them until we get customers asking for it, at which point we would expose it.

Reply:
I assume that all these marshaler gets used even without directly referencing the type. Is that correct?

Reply:
The other one does. This one does not. This one requires a direct (or indirect as I'm currently doing in the tests) reference. I'd prefer to keep both in this PR for completeness.

Reply:
Do we need to have it build into the product then? It looks like half-done feature. We should either not have it in the product at all; or we should have it in the product and be exposed as a proper public API. I do not think we want to be ever telling people to reference System.Private.CoreLib directly.

Reply:
I think this one used to be required in .NET Framework but its functionality was later integrated into the runtime. The fact that this marshaller isn't needed on the parameter but is required on the return value (which gives a "System.PlatformNotSupportedException: IDispatch and IDispatchEx are not supported" error when enumerating without it) is what prompts that idea in my mind.

Reply:
I'm fine exposing it via CoreFX in a ref assembly if that's what we want to do. Would you prefer putting it in a CustomMarshalers.dll (matching .NET Framework), or in System.Runtime.InteropServices (to avoid creating another ref-assembly).

Reply:
The question for me is whether we need to have it all.

Is all Office interop going to work when we do not have it all (ie not even in CoreLib)? If the answer to this is yet, I would not have it at all.

Reply:
I'm not sure. By the way, I found the portion of the code that is causing that error I mentioned above.

https://github.com/dotnet/coreclr/blob/71255bc05db24f078f334ead245898b817c8a06e/src/vm/mngstdinterfaces.cpp#L993-L995

If we implement this, then we don't need `EnumerableToDispatchMarshaler` at all (its functionality will be completely implemented within the runtime).

Reply:
By looking at the comment a little higher up in the file, it seems like the runtime should be using this marshaler automatically. I'll investigate a little more to see what we should be doing.

Reply:
> If we implement this, then we don't need EnumerableToDispatchMarshaler at all (its functionality will be completely implemented within the runtime).

That would be preferable, I think.

Reply:
It should be just about reverting https://github.com/dotnet/coreclr/pull/11865/ . That fix was made to avoid confusing error message - that was caused by these marshalers missing.

Reply:
That change is now in this PR. Now we don't need to expose this API to the user for it to be used correctly.